lib/decode.js

```diff
@@ -0,0 +1,292 @@
+}
+
+
+/**
+ * Create a new stripe object to track its content
+ *
+ * @param {DecodeContext} decodeContext - Context of the stream decoding
+ * @return {Object} new stripe informations
+ */
+function newStripe(decodeContext) {
+    return {
+        dataBuffer: Buffer.alloc(decodeContext.data.size, 0),
+        parityBuffer: Buffer.alloc(decodeContext.parity.size, 0),
+        waiting: decodeContext.sources,
+    };
+}
+
+
+/**
+ * Should we start encoding current stripe?
+ *
+ * @param {DecodeContext} decodeContext - Context of the stream decoding
+ * @return {boolean} true if stripe can be dispatched
+ */
+function shouldSendToEncoder(decodeContext) {
+    return (
+        // We have a stripe
+        decodeContext.stripe &&
+            // It is ready/filled entirely
+            decodeContext.stripe.waiting === 0 &&
+            // Decoder is ready
+            !decodeContext.waitEncoding());
+}
+
+
+/**
+ *Dispatch current stripe to decoder
+ *
+ * @param {DecodeContext} decodeContext - Context of the stream decoding
+ * @return {undefined}
+ */
+function sendToEncoder(decodeContext) {
+    decodeContext.ref();
+    decodeStep(decodeContext,
+               decodeContext.stripe.dataBuffer,
+               decodeContext.stripe.parityBuffer);
+    // eslint-disable-next-line no-param-reassign
+    decodeContext.stripe = newStripe(decodeContext);
+}
+
+
+/**
+ * Fill corresponding part of the stripe buffers with input data
+ *
+ * @param {DencodeContext} decodeContext - Context of the stream decoding
+ * @param {Number} istreamId - StreamId to read from
+ * @return {boolean} filled the buffer or not
+ */
+function fillStripeBuffers(decodeContext, istreamId) {
+    const size = decodeContext.stripeSize;
+    const bytes = decodeContext.istreams[istreamId].read(size);
+    // Not enough data to fill part of the stripe
+    if (bytes === null) {
+        return false;
+    }
+
+    if (istreamId < decodeContext.k) {// data
+        const copied = bytes.copy(decodeContext.stripe.dataBuffer,
+                                  istreamId * size);
+        assert.strictEqual(copied, bytes.length);
+    } else {
+        const copied = bytes.copy(decodeContext.stripe.parityBuffer,
+                                  (istreamId - decodeContext.k) * size);
+        assert.strictEqual(copied, bytes.length);
+    }
+
+    return true;
+}
+
+
+/**
+ * Bufferize stream stripes - Step 1
+ *
+ * @param {DencodeContext} decodeContext - Context of the stream decoding
+ * @param {Number} istreamId - StreamId to read from
+ * @return {undefined}
+ */
+function decodeBufferStep(decodeContext, istreamId) {
+    // Backpressure - avoids memory blow-up
+    if (decodeContext.waitReading()) {
+        return;
+    }
+
+    if (shouldSendToEncoder(decodeContext)) {
+        sendToEncoder(decodeContext);
+    }
+
+    if (!decodeContext.stripe) {
+        // eslint-disable-next-line no-param-reassign
+        decodeContext.stripe = newStripe(decodeContext);
```

Comment:
As mentioned in another comment, the code permits buffering a stripe while another is dispatched to the task pool for encoding. It might be a pointless optimization in our context though.

Reply:
I run a quick benchmark with (9, 3) RS code and 64K strip size for 
1. Current version
2. Modified version 1: use `allocUnsafe` 
3. Modified version 2: use `allocUnsafe` + allocate once `parityBuffer` and `stripe`

1. Current version:
```code
$ node benchmark.js
type,k,m,stripeSize,kill,latencyMs,bandwidthMBs
buffer,9,3,65536,0,2.5139519999999997,53389.15540153512
stream,9,3,65536,0,137.72205599999998,974.5554045460955
buffer,9,3,65536,1,1.693682,79246.14774202005
stream,9,3,65536,1,124.750999,1075.8853642526744
buffer,9,3,65536,2,1.796478,74711.61572810801
stream,9,3,65536,2,114.510336,1172.1018266857586
```

2. Modified version 1
```code
node benchmark.js
type,k,m,stripeSize,kill,latencyMs,bandwidthMBs
buffer,9,3,65536,0,2.3256319999999997,57712.38699845892
stream,9,3,65536,0,129.03241599999998,1040.1864753117543
buffer,9,3,65536,1,1.798682,74620.06847235921
stream,9,3,65536,1,107.535974,1248.1197594397574
buffer,9,3,65536,2,1.541282,87081.90584201983
stream,9,3,65536,2,91.457702,1467.5393221666557
```

3. Modified version 2
```code
node benchmark.js
type,k,m,stripeSize,kill,latencyMs,bandwidthMBs
buffer,9,3,65536,0,2.267064,59203.34582526122
stream,9,3,65536,0,72.72670699999999,1845.5087482511758
buffer,9,3,65536,1,1.6251579999999999,82587.52318236136
stream,9,3,65536,1,60.814752,2206.99369126754
buffer,9,3,65536,2,1.243058,107973.86284469429
stream,9,3,65536,2,48.930603,2743.0230933389485
```

=> We can double performance by allocating once

Reply:
Alright, will simplify the code then

Reply:
Fails for k+1 (bug open on 3rd party lib). Also note the benchmark didn't fail here because the streams are consumed directly. We are pushing buffers into a stream. Seems like they use a reference, and hence we can't reuse them for next stripe until they are consumed :/

I attached a quick test showing that. Hope I made a mistake, otherwise I don't think we can avoid allocating all the time, unless there is a clever way of knowing when the buffers are safe for reuse?

Reply:
The fault was mine in the 3rd party lib. You should be able to avoid unnecessary allocations once your PR lands there. Thanks!